example/libgcrypt/time_decrypt_oaep.c

```diff
@@ -0,0 +1,369 @@
+        switch (opt) {
+            case 'i':
+                in_file_name = optarg;
+                break;
+            case 'o':
+                out_file_name = optarg;
+                break;
+            case 'k':
+                key_file_name = optarg;
+                break;
+            case 'n':
+                sscanf(optarg, "%zi", &ciphertext_len);
+                break;
+            case 'h':
+                help(argv[0]);
+                exit(0);
+                break;
+            default:
+                fprintf(stderr, "Unknown option: %c\n", opt);
+                help(argv[0]);
+                exit(1);
+                break;
+        }
+    }
+
+    if (!in_file_name || !out_file_name || !key_file_name || !ciphertext_len) {
+        fprintf(stderr, "Missing parameters!\n");
+        help(argv[0]);
+        exit(1);
+    }
+
+    in_fd = open(in_file_name, O_RDONLY);
+    if (in_fd == -1) {
+        fprintf(stderr, "can't open input file %s\n", in_file_name);
+        goto err;
+    }
+
+    out_fd = open(out_file_name, O_WRONLY|O_TRUNC|O_CREAT, 0666);
+    if (out_fd == -1) {
+        fprintf(stderr, "can't open output file %s\n", out_file_name);
+        goto err;
+    }
+
+    fprintf(stderr, "malloc(plaintext)\n");
+    plaintext = malloc(ciphertext_len);
+    if (!plaintext)
+        goto err;
+
+    fprintf(stderr, "malloc(ciphertext)\n");
+    ciphertext = malloc(ciphertext_len);
+    if (!ciphertext)
+        goto err;
+
+    fprintf(stderr, "fopen()\n");
+    fp = fopen(key_file_name, "r");
+    if (!fp) {
+        fprintf(stderr, "Can't open key file %s\n", key_file_name);
+        goto err;
+    }
+
+    fprintf(stderr, "read_private_key()\n");
+    if ((pkey = read_private_key(fp))== NULL) {
+        fprintf(stderr, "Can't read key from file %s\n", key_file_name);
+        goto err;
+    }
+
+    fprintf(stderr, "fclose()\n");
+    if (fclose(fp) != 0)
+        goto err;
+    fp = NULL;
+
+    fprintf(stderr, "Decrypting ciphertexts...\n");
+
+    while ((r_ret = read(in_fd, ciphertext, ciphertext_len)) > 0) {
+        if (r_ret != ciphertext_len) {
+            fprintf(stderr, "read less data than expected (truncated file?)\n");
+            goto err;
+        }
+
+        gcry_sexp_t ciphertext_sexp;
+        gcry_sexp_t plaintext_sexp;
+        size_t erroff;
+
+        if (r_ret = gcry_sexp_build(&ciphertext_sexp, &erroff,
+                "(enc-val (flags oaep) (rsa (a %b)))",
+                ciphertext_len, ciphertext)) {
+            fprintf(stderr, "ciphertext s-expression construction failed\n");
+            fprintf(stderr, "error at pos %i\n", erroff);
+            fprintf(stderr, "error code: %i\n", r_ret);
+            fprintf(stderr, "failure: %s/%s\n",
+                    gcry_strsource(r_ret), gcry_strerror(r_ret));
+            goto err;
+        }
+
+        time_before = get_time_before();
+
+        r_ret = gcry_pk_decrypt(&plaintext_sexp, ciphertext_sexp, pkey);
```

Comment:
for OAEP we should verify that the decryption failed (similarly to the implicit rejection check that the decryption is 48 bytes long)

Reply:
Fixed.